File: ria/java/gof/src/main/java/com/gof/pattern/singleton/Settings6.java
package com.gof.pattern.singleton;

/**
 * enum 으로 싱글톤 패턴을 구현할 경우, 리플렉션에 안전함
 * enum 은 리플렉션에서 newInstance 를 막아놓음 -> 유일한 인스턴스 보장 가능
 * 단점 : enum 은 클래스 로딩시점부터 생성됨, 상속 불가능
 * enum 은 Serializable 을 구현하고 있음
 **/
public enum Settings6 {
    INSTANCE;
}

File: zeri/gof/src/main/kotlin/com/gof/pattern/singleton/javakotlin/Settings6.kt
package com.gof.pattern.singleton.javakotlin

/**
 * [멀티 쓰레드 환경에서 안전하게 구현하는 방법]
 * - static inner 클래스 사용하기
 * (권장하는 방법)
 */
class Settings6 private constructor() {

    companion object {
        private object SettingsHolder {
            val instance = Settings6()
        }

        fun getInstance(): Settings6 = SettingsHolder.instance
    }
}
